Fix #12062: Send emails to new or removed question submitter

laeeraa opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

This PR fixes #12062 .

This PR adds the functionality to send an email notification every time a user gets assigned or removed from the submit question allowlist
The email then looks as follows:
1. for added question submitter:  
Subject: _You have been invited to submit Oppia questions_ 

_Hi questionSuggestor,<br><br>This is to let you know that the Oppia team has added you as a contributor to submit question suggestions for use in lessons.<br><br>You can now start to submit questions in the <a href="https://www.oppia.org/contributor-dashboard">Contributor Dashboard</a>.<br><br>Thanks, and happy contributing!<br><br>Best wishes,<br>The Oppia Community_ 

3. for removed question submitter: 
Subject: _You have been unassigned as a question submitter_ 

_Hi questionSuggestor,<br><br>The Oppia team has removed you from the question submitter role. You won't be able to submit question suggestions any more, but you can still contribute questions through the <a href="https://www.oppia.org/contributor-dashboard">Contributor Dashboard</a>.<br><br>Thanks, and happy contributing!<br><br>Best wishes,<br>The Oppia Community_

Note "questionSuggesotr will be replaced by the username. 

Main changes:

- rename function send_email_to_new_reviewer to send_email_to_new_cd_user in email_manager.py
- update this function to also handle the added submitter event
- rename function send_email_to_new_reviewer to send_email_to_new_cd_user in email_manager.py
- update this function to also handle the removed submitter event
- rename specific model fields from reviewer_* to cd_user* since also submitter will be informed now
- update email_manager_test.py and add functions test_send_removed_translation_reviewer_email and test_send_assigned_question_submitter_email

since the naming of the category constants does not really fit I changed it in assets/constants.ts and all connected files: 

https://github.com/oppia/oppia/blob/4858ea094646a84f8cc31a5d323a1e9e58ca331b/assets/constants.ts#L6020-L6029

On my way, I dropped some of the voiceover code in 
- core/domain/email_manager.py
- core/domain/user_services.py
- core/controllers/contributor_dashboard_admin.py

## Essential Checklist

- [ x]  The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...",
followed by a short, clear summary of the changes.
- [x ]  I have followed the [[instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request)](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [ x]  The linter/Karma presubmit checks have passed on my local machine.
- [x ]  "Allow edits from maintainers" is checked. (See [[here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [x ]  My PR follows the [[style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide)](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [ x]  I have assigned the correct reviewers to this PR (or will leave a
comment with the phrase "@{{reviewer_username}} PTAL" if I don't have
permissions to assign reviewers directly).

## Proof that changes are correct

Since the development server cannot send any emails, end-to-end tests are not possible for me.
-  test_send_assigned_question_submitter_email tests functionality of send_email_to_new_cd_user
-  test_send_removed_question_submitter_email tests functionality of send_email_to_removed_cd_user
I have ensured with console prints that the right functions are called when pushing the buttons. 

Since this is my first issue to work on, I hope that the approach is correct and this is enough information to proof the functionality. Let me know if I should update or add anything else.

Updated Screen Recording of assigning and removing submitter/reviewer: (https://1drv.ms/u/s!AjYAChL9LPz-g9UVsVHzK8C4eYghQA?e=m4HSkb)
<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface. Please also include
videos/screenshots of the developer tools browser console, so that we can be
sure that there are no console errors.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the [["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR)](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [[Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers)](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

Commit:
add submitter category, adapt function send_email_to_new_contribution_reviewer to send_email_to_new contributior

Commit:
adapt Email text

Commit:
update/add email-manager tests

Commit:
send email to removed question submitter

Commit:
fix linter issues

Commit:
fix linter issues 2

Commit:
Merge branch 'develop' into send-emails-to-new-submitter

Commit:
remove debugging logging info

Commit:
fix some typos

Commit:
fix linter

Commit:
fix linter issues

Commit:
fix testCoverage

Commit:
Merge remote-tracking branch 'upstream/develop' into send-emails-to-new-submitter

Ash-2k3 commented:
Hey @laeeraa, the tests which are failing seems to be flaky tests so don't worry about them. Also is this PR ready for review ? If yes please follow this checkpoint to assign the required reviewers ` I have assigned the correct reviewers to this PR (or will leave a
comment with the phrase "@{{reviewer_username}} PTAL" if I don't have
permissions to assign reviewers directly).`

laeeraa commented:
@{{reviewer_username}} PTAL 
This PR is ready to review. Let me know if I should update anything.

laeeraa commented:
> Hey @laeeraa, the tests which are failing seems to be flaky tests so don't worry about them. Also is this PR ready for review ? If yes please follow this checkpoint to assign the required reviewers ` I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I don't have permissions to assign reviewers directly).`

Hey, thanks for getting in touch @Ash-2k3. Yes this PR is ready to review. I thought since the OppiaBot already assigned some reviewers, that I didn't need it. 
Is the comment I made below correct now?

Ash-2k3 commented:
Looks like you have misunderstood the checkpoint. In the curly braces, user name of the reviewer must come. For e.g (@Ash-2k3 PTAL.)

I have assigned the mentors this time, please make sure to assign the mentors from next time otherwise they might miss to review the PRs.

Ash-2k3 commented:
Hey @sahiljoster32 PTAL. (Was not able to search username in assigning section.)

seanlip commented:
Assigned @kevintab95 in lieu of @sahiljoster32 for lace-backend-reviewers.

vojtechjelinek reviewed (approved):
LGTM!

lkbhitesh07 reviewed (approved):
Thank you so much for the changes @laeeraa, LGTM for the code owner file.

kevintab95 reviewed (approved):
LGTM for codeowner files. Thanks @laeeraa!

sagangwee reviewed (approved):
@laeeraa LGTM, thanks1

seanlip commented on core/domain/email_manager.py:
```diff
@@ -2197,109 +2213,153 @@ def send_account_deletion_failed_email(user_id: str, user_email: str) -> None:
+    contribution_category_data = (
+        NEW_CONTRIBUTOR_EMAIL_DATA[contribution_category])
+    email_subject = 'You have been invited to %s Oppia %s' % (
+                    contribution_category_data['task'],
```
This and the next line should be indented by 4 from the previous line. The line after that should be indented at the same level as the line above (to match the scope of the '(' and ')').

laeeraa replied:
done: 

https://github.com/oppia/oppia/blob/36b6b47f846a94b261c017b0a3e3420fa86138b3/core/domain/email_manager.py#L2218-L2221

seanlip commented on core/domain/email_manager.py:
```diff
@@ -2197,109 +2213,153 @@ def send_account_deletion_failed_email(user_id: str, user_email: str) -> None:
+    if contribution_category not in REMOVED_CONTRIBUTOR_EMAIL_DATA:
+        raise Exception(
+            'Invalid contribution_category: %s' % contribution_category
+            )
```
Deindent by 4 to match opening paren. Ditto elsewhere.

laeeraa replied:
Done: 
https://github.com/oppia/oppia/blob/4858ea094646a84f8cc31a5d323a1e9e58ca331b/core/domain/email_manager.py#L2319-L2320

seanlip commented on core/domain/email_manager.py:
```diff
@@ -2197,109 +2213,153 @@ def send_account_deletion_failed_email(user_id: str, user_email: str) -> None:
+    if contribution_category in (
+        [constants.CONTRIBUTION_RIGHT_CATEGORY_SUBMIT_QUESTION]):
+        email_subject = 'You have been unassigned as a %s submitter' % (
+        contribution_category_data['contribution_category'])
```
Indent by 4, since '(' introduces a new scope.

Ditto elsewhere.

laeeraa replied:
done: 

https://github.com/oppia/oppia/blob/36b6b47f846a94b261c017b0a3e3420fa86138b3/core/domain/email_manager.py#L2323-L2329

seanlip commented on core/domain/email_manager.py:
```diff
@@ -61,34 +61,44 @@
 
 
-NEW_REVIEWER_EMAIL_DATA: Dict[str, Dict[str, str]] = {
+NEW_CONTRIBUTOR_EMAIL_DATA: Dict[str, Dict[str, str]] = {
```
Hi @laeeraa, sorry for not seeing this earlier, but I realized that there is an ambiguity in the chosen naming. We usually think of "contributors" as the folks who submit translations/questions, and their submissions are reviewed by reviewers. You are correct that, in a way, all of these folks are "contributors" because they use the contributor dashboard, but the existing association of the term "contributors" to the submitters makes this a bit confusing. So every time I read the code, I find myself having to step back and realize "contributors means everyone here, not just submitters" and I think this could cause confusion down the line too.

Would you be open to changing this to NEW_CD_USER_EMAIL_DATA, and adjusting similar nomenclature elsewhere (i.e. to CD_USER for cases which refer to both submitters and reviewers)? Here, "CD" means contributor dashboard. It is a bit awkward but I can't think of a better name that avoids the connotation issues I mentioned above.

Also, if it makes anything easier, feel free to drop any code relating to voiceover submissions. We don't currently support that, and aren't planning to do so going forward anyway. This is optional though, I only mention it in case it helps with figuring out the conceptual naming. 

Thanks!

laeeraa replied:
Ah okay! Sure, I will change it within the next two days:)

laeeraa replied:
![grafik](https://github.com/oppia/oppia/assets/92863142/ecc54eb7-dbfa-4acf-9b78-3337f9c66035)

I would also rename those constants in assets/constants.ts to: 

![grafik](https://github.com/oppia/oppia/assets/92863142/b8b4d16d-f201-4c6d-ac5c-6c13647cde73)

@seanlip would this be fine with you?

seanlip replied:
Yes, this is fine, let's just use RIGHTS instead of RIGHT here (otherwise it sounds like "correct" rather than their access permissions). Thanks!

laeeraa replied:
done with: 
https://github.com/oppia/oppia/blob/4858ea094646a84f8cc31a5d323a1e9e58ca331b/assets/constants.ts#L6020-L6029

Commit:
change naming from contributor to cd_user

Commit:
change naming from contributor to cd_user in tests

sahiljoster32 commented on core/domain/email_manager.py:
```diff
@@ -2197,109 +2213,153 @@ def send_account_deletion_failed_email(user_id: str, user_email: str) -> None:
+    if contribution_category in [
             constants.CONTRIBUTION_RIGHT_CATEGORY_REVIEW_TRANSLATION,
-            constants.CONTRIBUTION_RIGHT_CATEGORY_REVIEW_VOICEOVER]:
+            constants.CONTRIBUTION_RIGHT_CATEGORY_REVIEW_VOICEOVER
+            ]:
```
```suggestion
            constants.CONTRIBUTION_RIGHT_CATEGORY_REVIEW_VOICEOVER
        ]:
```

laeeraa replied:
done

sahiljoster32 commented on core/domain/email_manager.py:
```diff
@@ -2197,109 +2213,153 @@ def send_account_deletion_failed_email(user_id: str, user_email: str) -> None:
+            constants.CONTRIBUTION_RIGHT_CATEGORY_REVIEW_VOICEOVER,
+            constants.CONTRIBUTION_RIGHT_CATEGORY_REVIEW_VOICEOVER,
+            constants.CONTRIBUTION_RIGHT_CATEGORY_REVIEW_QUESTION
+            ]:
```
```suggestion
            constants.CONTRIBUTION_RIGHT_CATEGORY_REVIEW_TRANSLATION,
            constants.CONTRIBUTION_RIGHT_CATEGORY_REVIEW_VOICEOVER,
            constants.CONTRIBUTION_RIGHT_CATEGORY_REVIEW_VOICEOVER,
            constants.CONTRIBUTION_RIGHT_CATEGORY_REVIEW_QUESTION
        ]:
```

laeeraa replied:
done: 
https://github.com/oppia/oppia/blob/36b6b47f846a94b261c017b0a3e3420fa86138b3/core/domain/email_manager.py#L2251-L2254

sahiljoster32 reviewed (changes requested):
Some indentation issues, otherwise things looks good to me!

Commit:
change constants.CONTRIBUTION_RIGHTS_* to constants.CD_USER_RIGHTS_*

Commit:
remove voiceover category in email manager

Commit:
fix indent

Commit:
fix linter

Commit:
undo changes in contributor-dashboard-admin-page.component.ts

Commit:
fix linter

Commit:
add contributor page to test coverage

Commit:
fix linter

Commit:
fix linter

Commit:
order Not_fully_covered list

Commit:
fix not fully covered list

Commit:
fix not fully covered list

laeeraa commented:
Sorry that it took me so long... I was busy with exams.
I have now updated the naming of the variables and functions from Contributor_* to CD_USER_* and also fixed the indention issues. I have removed some of the voiceover code on my way.
I think it would be smart to open a new pull request to remove all unnecessary code of the voiceovers (if this is something that will actually not be used anymore).
I would also offer to take this on.

Commit:
merge develop

Commit:
fix constants names

Commit:
fix constants names

Commit:
fix category names

Commit:
remove contributor-dashboard from test-coverage exception

Commit:
fix linter

Commit:
fix linter

Commit:
fix variable name in frontend

Commit:
fix linter

laeeraa commented:
Hey! I will fix the remaining failing tests on Sunday.

Commit:
fix contributordashboard tests

Commit:
reset index.yaml

Commit:
fix frontendtest

Commit:
remove txtfile

laeeraa commented:
Hi! Could someone please help me to fix the remaining E2E test. I suppose it got something to do with me changing the name of the rightsvariables as stated in the description of this PR. 

To be honest, I don't really get what this test does (in core.tests.webdriverio_desktop.contributorAdminDashboard.js): 
`it('should allow translation coordinator to view dashboard',
    async function() {
      await users.login(TRANSLATION_COORDINATOR_EMAIL);
      await contributorDashboardAdminPage.get();

      await contributorDashboardAdminPage.navigateToTranslationSubmitterTab();
      await contributorDashboardAdminPage.waitForLoadingMessageToDisappear();
      await contributorDashboardAdminPage.expectNoStatsElement();

      await contributorDashboardAdminPage.switchLanguage('shqip (Albanian)');
      await contributorDashboardAdminPage.waitForLoadingMessageToDisappear();
      await contributorDashboardAdminPage.expectStatsElementCountToBe(2);
      await contributorDashboardAdminPage.expectStatsRowsAreExpanded();

      await contributorDashboardAdminPage.navigateToTranslationReviewerTab();
      await contributorDashboardAdminPage.waitForLoadingMessageToDisappear();
      await contributorDashboardAdminPage.expectStatsElementCountToBe(2);
      await contributorDashboardAdminPage.expectStatsRowsAreExpanded();

      await contributorDashboardAdminPage.switchLanguage('English');
      await contributorDashboardAdminPage.waitForLoadingMessageToDisappear();
      await contributorDashboardAdminPage.expectNoStatsElement(0);

      await users.logout();`

This is the test that fails saying: "Error: stats table is not visible"
If we are on the contributor-dashboard-admin page, there is no different tabs so I don't get why or where to we would change tabs and I also don't know which Stattable is expected. Is it the one at the very end of the contributor-dashboard-admin page? 

Thanks for your help!

kshitij01042002 commented:
Hi @laeeraa the tests that are failing is for the new feature we are implementing that is new contributor admin dashboard. It has different tabs and stats table. If you want to enable it please enable the feature flag `cd_admin_dashboard_new_ui`. You can follow the steps to enable the feature mentioned [here](https://github.com/oppia/oppia/wiki/Launching-new-features#changing-value-of-feature-flags).

Also I think you should keep the previous naming as you have made everything specifically for reviewers from what I can see but in contributor dashboard we have rights of reviewers as well as submitters, so keeping them as `translation` and `question` is fine. cc: @seanlip please let us know what are your thoughts on this.

Can you please keep the naming as previous and try again and see if the tests are still failing? Please let me know as well it'll help me debug and fix if something is broken.

Please let me know, if you still need any more clarification.

laeeraa commented:
> Hi @laeeraa the tests that are failing is for the new feature we are implementing that is new contributor admin dashboard. It has different tabs and stats table. If you want to enable it please enable the feature flag `cd_admin_dashboard_new_ui`. You can follow the steps to enable the feature mentioned [here](https://github.com/oppia/oppia/wiki/Launching-new-features#changing-value-of-feature-flags).
> 
> Also I think you should keep the previous naming as you have made everything specifically for reviewers from what I can see but in contributor dashboard we have rights of reviewers as well as submitters, so keeping them as `translation` and `question` is fine. cc: @seanlip please let us know what are your thoughts on this.
> 
> Can you please keep the naming as previous and try again and see if the tests are still failing? Please let me know as well it'll help me debug and fix if something is broken.
> 
> Please let me know, if you still need any more clarification.

Perfect thanks for the quick reply! I will look into it tomorrow morning and see if there is a quick fix to the remaining failing test and if not go back to the old variable naming. 

I changed it because the 'question' naming basically just gives the right to "review questions" while the 'submit_question' variable gives the right to submit questions. Same for the translation variable. So I thought it seems more intuitiv to actually name them that way as well.

Commit:
Merge remote-tracking branch 'upstream/develop' into send-emails-to-new-submitter

Commit:
revert naming of variable translation & question

Commit:
change order in EmailManagerTest

Commit:
get voiceover category back

Commit:
fix typo

laeeraa commented:
Hey, 
I have now changed back the variable naming and the E2E Test for the new feature page are passing now. However there is some new tests failing now: the additional Features Editor E2E Tests. Could someone please help me to face these failing tests since I am not quite sure if they are connected to my changes.

seanlip commented:
Hi @laeeraa -- I'm not sure either, it might be a flake. I've restarted the tests, let's see if it works this time. If not then please feel free to leave a reply in this thread and ask for help in GitHub Discussions if necessary (I'm travelling this week so will probably not be very responsive). Thanks!

seanlip commented:
@laeeraa Please could you respond to all comment threads as per the instructions in step 5 [here](https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve) and then tag the remaining reviewers following the instructions in the PR description? Please don't resolve comment threads; instead, reply "Done" to the ones you have addressed, and ensure that each comment thread has a reply. Thanks!

Commit:
Merge remote-tracking branch 'upstream/develop' into send-emails-to-new-submitter

laeeraa committed:
indention issue

laeeraa committed:
indention issues

laeeraa reviewed (commented):
fixed indention issues and updated the naming of the variables as discussed

laeeraa commented:
> @laeeraa Please could you respond to all comment threads as per the instructions in step 5 [here](https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve) and then tag the remaining reviewers following the instructions in the PR description? Please don't resolve comment threads; instead, reply "Done" to the ones you have addressed, and ensure that each comment thread has a reply. Thanks!

Hey @seanlip! Thanks for looking into this again. 
I have now unresolved all the conversations, replied to the requested changes with a permalink proof and rerequested a review. Sorry for not getting the workflow of this right away. I am still new to the open source world. I hope I've followed the correct workflow now. Let me know if I should add something else.
I have also left a discussion about the failing tests here: https://github.com/oppia/oppia/discussions/19123#discussion-5813276 but haven't gotten any reply yet.

seanlip commented:
Thanks @laeeraa. I think you still need to ensure that the right reviewers are assigned to your PR in the "Assignees" field. Please follow the instructions in the PR checklist (last bullet point) to do that. You will know that it is working correctly if the correct folks are listed in the "Assignees" column on the right. Thanks!

(Reviewers typically use the Assignees list to filter PRs assigned to them, so if you don't do this then your PR will likely never get reviewed. This is all documented in [step 5 here](https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve) -- see the very last bullet point.)

laeeraa commented:
@sahiljoster32 PTAL

laeeraa committed:
Merge branch 'develop' into send-emails-to-new-submitter

laeeraa commented:
@seanlip @DubeySandeep @Lawful2002 PTAL

laeeraa commented:
> Thanks @laeeraa. I think you still need to ensure that the right reviewers are assigned to your PR in the "Assignees" field. Please follow the instructions in the PR checklist (last bullet point) to do that. You will know that it is working correctly if the correct folks are listed in the "Assignees" column on the right. Thanks!
> 
> (Reviewers typically use the Assignees list to filter PRs assigned to them, so if you don't do this then your PR will likely never get reviewed. This is all documented in [step 5 here](https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve) -- see the very last bullet point.)

All right, I hope this is correct now. I have mentioned the remaining reviewers in comments since I do not have the rights to assign them myself. Should they be assigned automatically now? Am I missing some remaining review comment to address? 
I am so sorry for the confusion.

seanlip commented on core/domain/email_manager.py:
```diff
@@ -2197,132 +2192,166 @@ def send_account_deletion_failed_email(user_id: str, user_email: str) -> None:
-            constants.CONTRIBUTION_RIGHT_CATEGORY_REVIEW_TRANSLATION,
-            constants.CONTRIBUTION_RIGHT_CATEGORY_REVIEW_VOICEOVER]:
+    if category in [
+            constants.CD_USER_RIGHTS_CATEGORY_REVIEW_TRANSLATION,
```
Deindent this by 4, or move it to the end of the previous line / replace it with an equality.

laeeraa replied:
Done! Thanks for helping me with the contribution workflow.

seanlip reviewed (approved):
Thanks @laeeraa, this looks pretty good to me for my codeowner files. I just had one small comment.

Also thanks for fixing the reviewers issue! Oppiabot actually didn't assign properly that time but I think that's an issue with Oppiabot since we have one reviewer on this PR who isn't reviewing currently (and it couldn't handle that case). I've fixed this manually but thanks for following the instructions!

seanlip commented:
@DubeySandeep @Lawful2002 PTAL, thanks.

Lawful2002 reviewed (approved):
Thanks @laeeraa, LGTM for codeowner files!

laeeraa committed:
update indention & if-query

Commit:
Merge branch 'send-emails-to-new-submitter' of https://github.com/laeeraa/oppia into send-emails-to-new-submitter

Commit:
Merge remote-tracking branch 'upstream/develop' into send-emails-to-new-submitter

seanlip committed:
Merge branch 'develop' into send-emails-to-new-submitter

seanlip commented:
The only remaining CI failure is a flake so I think we can merge this.
